File: rust/src/leetcode/Q5_longest_palindromic_substring.rs
/*
 * @lc app=leetcode id=5 lang=rust
 *
 * [5] Longest Palindromic Substring
 */

// @lc code=start
impl Solution {
    pub fn longest_palindrome(s: String) -> String {
        let mut start: usize = 0;
        let mut length: usize = 0;

        let word: Vec<char> = s.chars().collect();
        let mut dp: Vec<Vec<bool>> = vec![vec![false; word.len()]; word.len()];

        for i in (0..word.len()).rev() {
            for j in i..word.len() {
                if word[i] == word[j] && (j - i <= 2 || dp[i + 1][j - 1]) {
                    dp[i][j] = true;

                    if j - i + 1 > length {
                        start = i;
                        length = j - i + 1;
                    }
                }
            }
        }

        word[start..start + length].iter().collect()
    }
}
// @lc code=end

File: python/Q5_longest_palindromic_substring.py
#
# @lc app=leetcode id=5 lang=python3
#
# [5] Longest Palindromic Substring
#


# @lc code=start
class Solution:
    def longestPalindrome(self, s: str) -> str:
        t: str = "@#" + "#".join(s) + "#$"
        p: list[int] = [0] * len(t)

        center: int = 0
        right: int = 0

        result: int = 0
        length: int = 0

        for i in range(1, len(t) - 1):
            if i <= right:
                p[i] = right - i if right - i < p[2 * center - i] else p[2 * center - i]

            while t[i - p[i] - 1] == t[i + p[i] + 1]:
                p[i] += 1

            if i + p[i] > right:
                right = i + p[i]
                center = i

            if p[i] > length:
                length = p[i]
                result = i

        palindrome_in_t: str = t[result - length : result + length + 1]
        return palindrome_in_t.replace("#", "").replace("@", "").replace("$", "")


# @lc code=end
